packages/@react-aria/interactions/src/useMove.ts

```diff
@@ -0,0 +1,220 @@
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+import {disableTextSelection, restoreTextSelection}  from './textSelection';
+import {HTMLAttributes, useMemo, useRef} from 'react';
+
+export interface BaseMoveEvent {
+  pointerType: 'mouse' | 'pen' | 'touch' | 'keyboard'
+}
+
+export interface MoveStartEvent extends BaseMoveEvent{
+  type: 'movestart'
+}
+
+export interface MoveMoveEvent extends BaseMoveEvent{
+  type: 'move',
+  deltaX: number,
+  deltaY: number
+}
+
+export interface MoveEndEvent extends BaseMoveEvent{
+  type: 'moveend'
+}
+
+export type MoveEvent = MoveStartEvent | MoveMoveEvent | MoveEndEvent;
+
+export interface MoveProps {
+  onMoveStart?: (e: MoveStartEvent) => void,
+  onMove: (e: MoveMoveEvent) => void,
+  onMoveEnd?: (e: MoveEndEvent) => void
+}
+
+export function useMove(props: MoveProps): HTMLAttributes<HTMLElement> {
+  let {onMoveStart, onMove, onMoveEnd} = props;
+
+  let state = useRef<{
+    didMove: boolean,
+    lastPosition: {pageX: number, pageY: number} | null,
+    id: number | null
+  }>({didMove: false, lastPosition: null, id: null});
+
+  let moveProps = useMemo(() => {
+    let moveProps: HTMLAttributes<HTMLElement> = {};
+
+    let start = () => {
+      disableTextSelection();
+      state.current.didMove = false;
+    };
+    let move = (pointerType: BaseMoveEvent['pointerType'], deltaX: number, deltaY: number) => {
+      if (!state.current.didMove) {
+        state.current.didMove = true;
+        onMoveStart?.({
+          type: 'movestart',
+          pointerType
+        });
+      }
+      onMove({
+        type: 'move',
+        pointerType,
+        deltaX: deltaX,
+        deltaY: deltaY
+      });
+    };
+    let end = (pointerType: BaseMoveEvent['pointerType']) => {
+      restoreTextSelection();
+      if (state.current.didMove) {
+        onMoveEnd?.({
+          type: 'moveend',
+          pointerType
+        });
+      }
+    };
+
+    if (typeof PointerEvent === 'undefined') {
+      let onMouseMove = (e: MouseEvent) => {
+        if (e.button === 0) {
+          move('mouse', e.pageX - state.current.lastPosition.pageX, e.pageY - state.current.lastPosition.pageY);
+          state.current.lastPosition = {pageX: e.pageX, pageY: e.pageY};
+        }
+      };
+      let onMouseUp = (e: MouseEvent) => {
+        if (e.button === 0) {
+          end('mouse');
+          window.removeEventListener('mousemove', onMouseMove, false);
+          window.removeEventListener('mouseup', onMouseUp, false);
```

Comment:
We should probably move to a disposable pattern for these global listeners (kinda like what's in usePress), and make sure these also get removed on unmount (in case the component unmounts while being interacted with).

Reply:
Done